### ISSUE TITLE
Multi-state agents

### ISSUE THREAD

**Issue by arctangent** [arctangent/simulated_agency#8]
At the moment, we have assumed that an agent can only have one "state". This is fairly standard for agent-based models.

However, to allow agents to formulate "plans" it may make sense to support a "state stack".

Example:

- Our agent is an ant
- The ant is at it's nest
- The ant notices that the nest is low on food
- The ant formulates the following plan (of states): [ FIND_FOOD, RETURN_TO_NEST ]
- The ant executes the FIND_FOOD state plan
- The ant then executes the RETURN_TO_NEST state plan

Allowing the ant to formulate a multi-stage state stack allows for more subtle behaviour (than just assuming all ants with food will want to return to the nest).

Perhaps the above is a trivial example (as we probably *do* want ants with food to go back to their nest!), but imagine if the ant is interrupted by an enemy ant either before or after it has collected some food. The state stack could then be modified in one of the following ways and still represent a sensible behaviour pattern:

- [ AVOID_ENEMY, FIND_FOOD, RETURN_TO_NEST ]
- [ AVOID_ENEMY, RETURN_TO_NEST ]

To summarise, allowing agents to formulate a stack of states is equivalent to allowing them to formulate multi-stage plans. It is a way to represent some kind of persistent knowledge about the world and what their larger goals are, without the need to store variables or flags.

**Comment by arctangent** [comment]
Initial implementation completed here: https://github.com/arctangent/simulated_agency/commit/9209ae62926ac125349e7b8dd8fb9c51155748c5
